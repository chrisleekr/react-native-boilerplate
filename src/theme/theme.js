import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
