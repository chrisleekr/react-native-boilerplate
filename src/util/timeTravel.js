/* eslint-disable no-undef */
// timeTravel.js: https://benjaminjohnson.me/testing-animations-in-react-native
//  This is a helper file to test animation.
import MockDate from 'mockdate';

const FRAME_TIME = 10;

const advanceOneFrame = () => {
  const now = Date.now();
  MockDate.set(new Date(now + FRAME_TIME));
  jest.advanceTimersByTime(FRAME_TIME);
};

const timeTravel = msToAdvance => {
  const numberOfFramesToRun = msToAdvance / FRAME_TIME;
  let framesElapsed = 0;

  // Step through each of the frames until we've ran them all
  while (framesElapsed < numberOfFramesToRun) {
    advanceOneFrame();
    framesElapsed++;
  }
};

export default timeTravel;

// timeTravel.js
export const setupTimeTravel = () => {
  MockDate.set(0);
  jest.useFakeTimers();
};
