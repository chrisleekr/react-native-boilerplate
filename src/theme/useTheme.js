/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
