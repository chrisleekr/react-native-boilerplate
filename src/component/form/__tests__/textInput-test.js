import 'react-native';
import React from 'react';

import {
  cleanup,
  fireEvent,
  render,
  within
} from '@testing-library/react-native';

import timeTravel, {setupTimeTravel} from '../../../util/timeTravel';
import {ThemeProvider} from '../../../theme/useTheme';

import TextInput from '../textInput';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn()
}));
import {useState} from 'react';

let rendered;
let component;
let element;

const onBlur = jest.fn();
const onChangeEmail = jest.fn();
const setFieldTouched = jest.fn();
const mockSetState = jest.fn();

describe('TextInput', () => {
  afterEach(cleanup);
  beforeEach(setupTimeTravel);

  beforeEach(() => {
    useState.mockImplementation(init => [init, mockSetState]);

    rendered = render(
      <ThemeProvider>
        <TextInput
          label={'My Label'}
          textInputStyle={{
            fontSize: 15
          }}
          field={{
            name: 'email-address',
            onBlur,
            onChange: jest.fn().mockImplementation(name => {
              switch (name) {
                case 'email-address':
                  return onChangeEmail;
              }
            }),
            value: 'Initial text'
          }}
          form={{
            errors: {},
            touched: {},
            setFieldTouched
          }}
          keyboardType="email-address"
        />
      </ThemeProvider>
    );

    component = rendered.getByTestId('text-input');
  });

  it('matches snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('does not show errors', () => {
    expect(within(component).queryByTestId('errors')).toBeNull();
  });

  describe('onChangeText', () => {
    beforeEach(() => {
      element = within(component).getByTestId('input');

      fireEvent(element, 'onChangeText', 'New Text');
    });

    it('should trigger onChangeEmail', () => {
      expect(onChangeEmail).toHaveBeenCalledWith('New Text');
    });
  });

  describe('onFocus', () => {
    beforeEach(() => {
      element = within(component).getByTestId('input');

      fireEvent(element, 'onFocus');
    });

    it('should set state for isFocused', () => {
      expect(mockSetState).toHaveBeenCalledWith(true);
    });
  });

  describe('onBlur', () => {
    beforeEach(() => {
      element = within(component).getByTestId('input');
      fireEvent(element, 'onBlur');
    });

    it('should set mockSetState as false', () => {
      expect(mockSetState).toHaveBeenCalledWith(false);
    });

    it('should trigger setFieldTouched', () => {
      expect(setFieldTouched).toHaveBeenCalledWith('email-address');
    });

    it('should trigger onBlur', () => {
      expect(onBlur).toHaveBeenCalledWith('email-address');
    });
  });

  describe('error text', () => {
    beforeEach(() => {
      rendered = render(
        <ThemeProvider>
          <TextInput
            label={'My Label'}
            textInputStyle={{
              fontSize: 15
            }}
            field={{
              name: 'email-address',
              onBlur,
              onChange: jest.fn().mockImplementation(name => {
                switch (name) {
                  case 'email-address':
                    return onChangeEmail;
                }
              }),
              value: 'Initial text'
            }}
            form={{
              errors: {
                'email-address': 'Please enter a valid email address.'
              },
              touched: {
                'email-address': true
              },
              setFieldTouched
            }}
            keyboardType="email-address"
          />
        </ThemeProvider>
      );

      component = rendered.getByTestId('text-input');
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('shows errors', () => {
      expect(within(component).queryByTestId('errors')).toBeTruthy();
    });
  });

  describe('Animated.timing', () => {
    beforeEach(() => {
      // Use actual useState
      useState.mockImplementation(jest.requireActual('react').useState);

      rendered = render(
        <ThemeProvider>
          <TextInput
            label={'My Label'}
            textInputStyle={{
              fontSize: 15
            }}
            field={{
              name: 'email-address',
              onBlur,
              onChange: jest.fn().mockImplementation(name => {
                switch (name) {
                  case 'email-address':
                    return onChangeEmail;
                }
              }),
              value: ''
            }}
            form={{
              errors: {
                'email-address': 'Please enter a valid email address.'
              },
              touched: {
                'email-address': true
              },
              setFieldTouched
            }}
            keyboardType="email-address"
          />
        </ThemeProvider>
      );

      component = rendered.getByTestId('text-input');
      element = within(component).getByTestId('input');
      fireEvent(element, 'onFocus');
      timeTravel(200);
      fireEvent(element, 'onBlur');
      timeTravel(200);
    });

    it('shows animated label', () => {
      expect(within(component).queryByTestId('label')).toBeTruthy();
    });
  });
});
