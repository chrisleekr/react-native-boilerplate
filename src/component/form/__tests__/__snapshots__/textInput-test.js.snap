// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput error text matches snapshot 1`] = `
<View
  style={
    Object {
      "paddingTop": 18,
    }
  }
  testID="text-input"
>
  <Text
    collapsable={false}
    style={
      Object {
        "color": "#525E75",
        "fontSize": 14,
        "left": 0,
        "position": "absolute",
        "top": 18,
      }
    }
    testID="label"
  >
    My Label
  </Text>
  <TextInput
    keyboardType="email-address"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    style={
      Array [
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "height": 44,
          "width": "100%",
        },
        Object {
          "fontSize": 15,
        },
        Object {
          "color": "#000000",
        },
        Object {
          "borderColor": "red",
        },
      ]
    }
    testID="input"
    value="Initial text"
  />
  <Text
    style={
      Object {
        "color": "red",
        "fontSize": 10,
        "marginBottom": 10,
      }
    }
    testID="errors"
  >
    Please enter a valid email address.
  </Text>
</View>
`;

exports[`TextInput matches snapshot 1`] = `
<View
  style={
    Object {
      "paddingTop": 18,
    }
  }
  testID="text-input"
>
  <Text
    collapsable={false}
    style={
      Object {
        "color": "#525E75",
        "fontSize": 14,
        "left": 0,
        "position": "absolute",
        "top": 18,
      }
    }
    testID="label"
  >
    My Label
  </Text>
  <TextInput
    keyboardType="email-address"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    style={
      Array [
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "height": 44,
          "width": "100%",
        },
        Object {
          "fontSize": 15,
        },
        Object {
          "color": "#000000",
        },
        undefined,
      ]
    }
    testID="input"
    value="Initial text"
  />
</View>
`;
